src/Bicep.Core/Workspaces/ParamSourceFileFactory.cs

```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using Bicep.Core.Parsing;
+using Bicep.Core.Text;
+
+namespace Bicep.Core.Workspaces
+{
+    public static class ParamSourceFileFactory
```

Comment:
I think we can modify the `SourceFileFactory` instead. `BicepParamFile` implements `ISourceFile` so it should be simple to incorporate it into the existing factory.

Reply:
I am a bit uncertain on how to go about properly doing this. Should I change the return value of  following function to ```ISourceFile``` and initialize the correct file object based on the extension in the ```Uri```?

https://github.com/Azure/bicep/blob/476668e54544f0c1a0c746fcffd2066c554c1c3d/src/Bicep.Core/Workspaces/SourceFileFactory.cs#L38

Reply:
We can move `public static BicepParamFile CreateBicepParamFile(Uri fileUri, string fileContents)` to `SourceFileFactory` and update `CreateSourceFile` accordingly. If the file extension of `fileUri` is `.bicepparam`, call `BicepParamFile` and return the result.

Reply:
So I made the change you've suggested, however, it looks like I need to type cast the file type I get from ```SourceFileFactory``` as follows
https://github.com/Azure/bicep/blob/8ea39ffd77ffa690382b38b175846b76e2c1e51b/src/Bicep.Cli/Services/CompilationService.cs#L108

The only way around this I can think of is to change the type of ```bicepParamFile``` parameter in ```ParamsSemanticModel``` to ```ISourceFile```. I don't think we are doing this in ```SemanticModel``` for Bicep so I am not sure if it would lead to issues later down the line if we decide to add things to ```BiceParamFile```. Any thoughts on this?

Reply:
Ah I see. In that case we can just call `SourceFileFactory.CreateParamsFile` and leave `CreateSourceFile` unchanged.